Parallelize `label clone`

heaths opened the pull request:
Greatly increases `label clone` speed, added in #5441, using similar algorithm to `status`.

heaths committed:
Parallelize `label clone`

Greatly increases `label clone` speed using similar alogirithm to `status`.

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

heaths commented:
Contains #5675 but could rebase if that gets merged first.

heaths commented on pkg/cmd/label/clone.go:
```diff
@@ -105,32 +107,67 @@ func cloneRun(opts *cloneOptions) error {
-				return
-			}
-		} else {
+	doneCh := make(chan struct{})
+	go func() {
+		for range created {
 			successCount++
 		}
+		close(doneCh)
+	}()
```
@samcoe this is the same pattern you added to `status` but I think there's a potential deadlock issue here. In trying to figure out why the mac test failed (it's still not clear since that's the only API that would return an error in the `errgroup.Group`), I found this would occasionally deadlock in an error case. I think we also need to select `ctx.Done()` and/or check if the channel was closed (might actually be redundant). I'm pushing another commit that shows this. See what you think, and I'd be happy to add another commit to fix `status` or open another PR.

heaths replied:
I appreciate the Go axiom "share by communicating", but for just incrementing a counter perhaps this is a bit convoluted? I still think the new pattern might be good to use for `status` but what about just using a mutex here or `atomic.AddUint32()` from `sync/atomic`? The `errgroup` support would still be valuable - just thinking about eliminating the separate goroutine to handling incrementing the counter.

heaths committed:
Fix deadlock issue

Also adds a couple more tests to cover corner cases.

heaths committed:
Fix intermittent error

Reduce "create error" scenario to a single bug to reduce the change of an intermittent error where there are not enough endpoint matches when verified.

mislav commented on pkg/cmd/label/clone.go:
```diff
@@ -105,32 +107,73 @@ func cloneRun(opts *cloneOptions) error {
+					createErr := createLabel(client, destination, &l)
+					if createErr != nil {
+						if !errors.Is(createErr, errLabelAlreadyExists) {
+							err := createErr
```
Nit: why assign before returning? It could just be `return createErr`

heaths replied:
Using named returns here so I need to assign `err`.

mislav replied:
Maybe I'm missing something, but the worker func that this is in doesn't have named returns. If you simply return an error object here without assigning, that would propagate through the errgroup and get returned `err = wg.Wait()` in the parent function, which does have named returns.

Finally, `err := ...` (colon-assignment) would shadow a named returns assignment, since it will always create a new variable for the current scope.

heaths replied:
Sorry, you're right. Wasn't thinking about that this was in the goroutine.

mislav commented on pkg/cmd/label/list.go:
```diff
@@ -76,8 +76,8 @@ func newCmdList(f *cmdutil.Factory, runF func(*listOptions) error) *cobra.Comman
 	cmd.Flags().IntVarP(&opts.Query.Limit, "limit", "L", 30, "Maximum number of labels to fetch")
 	cmd.Flags().StringVarP(&opts.Query.Query, "search", "S", "", "Search label names and descriptions")
-	cmdutil.StringEnumFlag(cmd, &opts.Query.Order, "order", "", "asc", []string{"asc", "desc"}, "Order of labels returned")
-	cmdutil.StringEnumFlag(cmd, &opts.Query.Sort, "sort", "", "created", []string{"created", "name"}, "Sort fetched labels")
```
Please rebase to take out the unrelated fix!

heaths replied:
Was planning on it. :)

mislav commented on pkg/cmd/label/clone.go:
```diff
@@ -105,32 +107,73 @@ func cloneRun(opts *cloneOptions) error {
-		Force: force,
+	workers := 10
+	toCreate := make(chan createOptions)
+	created := make(chan uint8)
```
Nit: since the value doesn't matter, an idiomatic approach in Go would be to use an empty struct as value, just like you do with `doneCh`

heaths replied:
Is it even worth it, actually? See [my comment](https://github.com/cli/cli/pull/5687#discussion_r878746235). What about just an atomic increment in this case? But this pattern could fix `gh status` which has the same potential deadlock I believe.

mislav replied:
Idiomatic approach vs. non-idiomatic is always worth it. But you're right that atomic increment might be even simpler. Or, have workers do the increment themselves, where the increment is guarded by a mutex.

mislav replied:
What is the deadlock case in `gh status`?

heaths replied:
Exactly - was just thinking the workers would use `atomic.AddUint32`.

This is the same pattern that @samcoe used for `gh status` which works, except in some error cases the `errgroup.Group` (also: better/easier to use `errgroup.WithContext` as I did here since it handles the cancellation for you) terminates but the goroutine that is waiting to collect information from the workers continues looping and never exits. In tests this would time out. In various CIs for this PR, it repro'd a couple times, and locally it repro'd quite a few times. I think the anonymous goroutine polling for mentions in `gh status` needs to also check `<-ctx.Done()` to exit when the `errgroup` exits.

If no error happens it should all work fine. I was going to add a specific error case for `gh status` and see if I can repro the deadlock and fix it separately if you think this pattern - as it is now in the PR (I'll use `atomic` in this 1 case) - is fine.

mislav commented on pkg/cmd/label/clone.go:
```diff
@@ -105,32 +107,73 @@ func cloneRun(opts *cloneOptions) error {
+		for {
+			select {
+			case <-ctx.Done():
+				break loop
```
Could you `return` instead of `break loop`? That would alleviate the need for a label

heaths replied:
I could. The reason was that I need to close the channel in a couple cases - potentially more clean up later, but future revisions could worry about that. Was just trying to reduce code duplication.

mislav reviewed (approved):
This looks good! I've left a few nitpicks as comments.

Yes, in theory the fan-in approach could be simplified (I also find incrementing the counter within a goroutine a bit convoluted), but I find no fault with your current approach and I don't think it's worth refactoring it too much.

For the future: since we repeat the pattern of “N workers handling a queue of X jobs” throughout the codebase and need to collect the results of those jobs + errors, it might be worth it to come up with a generic pipeline implementation to handle them all that fits our needs. That way we could be reusing the same code instead of having to roll it by hand each time.

heaths committed:
Resolve PR feedback
